### PR TITLE
[Streaming][bugfix] handle TLS signalisation when TLS is disabled on client side

### LINKED ISSUE #9474: [streaming] weird errors when watching services with streaming on Consul 1.9.1
#### Overview of the Issue

We upgraded servers from 1.8.6 to 1.9.1 with streaming enabled by using configuration:

```json
{
  "rpc": {
    "enable_streaming": true
  },
  "use_streaming_backend": true
}
```
on both servers and clients.

While all of this seemed to work in our test environment, in our preproduction, it seems to fail on all the DCs of our preproduction.

When trying such curl command on 1 agent for instance:
```
curl "http://localhost:8500/v1/health/service/consul-info-html?wait=1s&index=1&stale"
< HTTP/1.1 500 Internal Server Error
< Vary: Accept-Encoding
< X-Consul-Default-Acl-Policy: allow
< Date: Mon, 28 Dec 2020 22:17:27 GMT
< Content-Length: 16
< Content-Type: text/plain; charset=utf-8

context canceled
```

As an alternative, we also sometime have HTTP 500 with message:
```
rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection closed
```

which generate this kind of logs:

```
2020-12-28T22:22:39.181Z [ERROR] agent: subscribe call failed: err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection closed" topic=ServiceHealth key=consul-info-html failure_count=1
2020-12-28T22:22:39.181Z [ERROR] agent: subscribe call failed: err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection closed" topic=ServiceHealth key=consul-info-html failure_count=2
2020-12-28T22:22:39.181Z [ERROR] agent.http: Request error: method=GET url=/v1/health/service/consul-info-html?wait=30s&index=1&stale from=192.168.75.139:55754 error="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection closed"
```

Removing the `index=1` return the correct result:

```
curl "http://localhost:8500/v1/health/service/consul-info-html?wait=1s&stale" -i
HTTP/1.1 200 OK
Content-Type: application/json
Vary: Accept-Encoding
X-Consul-Default-Acl-Policy: allow
X-Consul-Index: 2070313186
X-Consul-Knownleader: true
X-Consul-Lastcontact: 90
Date: Mon, 28 Dec 2020 22:33:34 GMT
Transfer-Encoding: chunked

[{"Node":{"ID":"671c839d-aae4-9f9e-1b8e-bc57558444b1" ...]
```
Using any value for index (including the current index of service fails the same way (ie, in this example, using 2070313186 also fails)

The failure is the same on our 3 DCs, error does not happen if `"use_streaming_backend": false` is set on config of agent.

For all agents, /v1/agent/self => says both streaming is enabled.

Also tried to restart sequentially all servers + client, nothing works always in the same state.

### PR DESCRIPTION
**Pull request by pierresouchay**
This will fix #9474

`wrapper(server.Datacenter, conn)` could return the same connection. In such case, Consul client were incorrectly marking connection as RPCTLS while the connection was not encrypted.

This cause messages like:
```
[ERROR] agent.server.rpc: failed to read byte: conn=from=10.236.195.48:51572 error="tls: first record does not look like a TLS handshake"
```

The bug can be highlighted very easily by running Consul in dev mode with this config:
```
consul agent -dev -hcl 'rpc{enable_streaming=true}' -hcl use_streaming_backend=true -hcl 'ca_file="/etc/consul/ca.pem", cert_file="/etc/consul/cert.pem", key_file="/etc/consul/key.pem"'
```
Then, the bug is triggered with:
```
curl http://localhost:8500/v1/health/service/consul?index=2&stale
```

After this fix, such configuration works well.

### CODE REVIEW

File: agent/grpc/client.go [hunk 1]
```diff
@@ -94,19 +94,20 @@ func newDialer(servers ServerLocator, wrapper TLSWrapper) func(context.Context,
...
 			}
 
-			// Switch the connection into TLS mode
-			if _, err := conn.Write([]byte{byte(pool.RPCTLS)}); err != nil {
-				conn.Close()
-				return nil, err
-			}
-
-			// Wrap the connection in a TLS client
+			// Wrap the connection in a TLS client, return same conn if TLS disabled
```
Comment (dnephin) [inline agent/grpc/client.go:97]: Is TLS is disabled shouldn't this entire block be skipped because of the `server.UseTLS` check above?

**Review by dnephin — COMMENTED** [review]
Thank you for the PR! I was hoping to have a look at the bug report soon. This definitely seems like the right place for the fix.

I'm going to look into this fix some more tomorrow and see why the test is failing.

It seems a bit strange me to that the wrapper could return the same connection unmodified. I'm wondering if we don't need to send the RPCTLS byte at all from here, because that wrapping happens elsewhere, or if we should be checking something other than `server.UseTLS` or if there is something else I don't understand here.

File: agent/grpc/client.go (hunk 1 above)
Comment (pierresouchay) [reply agent/grpc/client.go:97]: Agree, I made a few weird tests, so, not sure about what is really happening.
The weird stuff is:
in dev mode with:
```
consul agent -dev -hcl 'rpc{enable_streaming=true}' -hcl use_streaming_backend=true -hcl 'ca_file="/etc/consul/ca.pem", cert_file="/etc/consul/cert.pem", key_file="/etc/consul/key.pem"'
```
=> for some reason it fixes the issue. But not sure yet about the why... trying to understand better.
I suspect that when running the streaming config on the server, the data is just proxified internally but is not encrypted... this part is a bit new to me, still looking

Comment (pierresouchay) [reply agent/grpc/client.go:97]: @dnephin Ok, found it, this is linked to `verify_outgoing: true|false`

* With verify_outgoing = true => works
* with verify_outgoing = false => fails

To test, use master + launch client with:

```
consul agent -dev -hcl 'rpc{enable_streaming=true}' -hcl use_streaming_backend=true -hcl 'telemetry{prometheus_retention_time = "192h", disable_compat_1.9=true}' -hcl 'ca_file="/etc/consul/ca.pem", cert_file="/etc/consul/cert.pem", key_file="/etc/consul/key.pem", verify_outgoing=false'
```
a call to 
`curl -i "http://localhost:8500/v1/health/service/consul?wait=10m&index=12&pretty=true&stale"`

fails with `verify_outgoing=false`, succeeds with `verify_outgoing=true`

Comment (pierresouchay) [reply agent/grpc/client.go:98]: @dnephin Found the issue, probably fixed properly in the next patch, but I had to do a bit more refactoring (essentially the tests)
Now works with both TLS and no TLS enabled

**Comment by dnephin** [comment]
Thank you for tracking down the bug! @kyhavlov and I were looking at this, and I think it should be possible to fix without changing the signature of `OutgoingRPCWrapper`.

As you found, the source of the problem is `verify_outgoing=false`. I think that's because this line: https://github.com/hashicorp/consul/blob/19baf4bc257d9c5379e0fb3df5c9ebfd9b6ec82c/tlsutil/config.go#L769

Which gets here:
https://github.com/hashicorp/consul/blob/19baf4bc257d9c5379e0fb3df5c9ebfd9b6ec82c/tlsutil/config.go#L558-L561

`verify_outgoing` seems like the wrong thing to be checking in this case. What we want to confirm is that:
1. the server will accept TLS (already handled by `newDialer` with `if server.UseTLS`)
2. the client is capable of using TLS (has the CA config)

`verify_outgoing` is neither of those, it is a way of saying "*must* use TLS). Even if `verify_outgoing` is false, I think we should opt to use TLS when the server supports it.

I think we can accomplish that by changing `OutgoingRPCWrapper` to call `MutualTLSCapable` instead of `UseTLS` on the line I referenced above.

We took a quick look at the two other callers of `OutgoingRPCWrapper` and both already explicitly call `UseTLS` themselves, so there should be no impact on the existing callers. `MutualTLSCapable` seems like it is the right check here. 

`UseTLS` should probably be renamed to something more appropriate for what it does, but we don't need to change that to fix the bug.

If we make that change I guess we also need to handle the case where `MutualTLSCapable()` returns false, but `verify_outgoing=true`. I'm not sure if that gets handled during config loading, or if that is something we'll need to check here. It seems like it should be possible to handle that during config loading, but I haven't looked to see if we do.

**Comment by pierresouchay** [comment]
@dnephin Thank you for looking. Well, to me the big problem is that `getAreaForPeerDatacenterUseTLS(dc)` called by `c.UseTLS(dc)` offers no guarantee that it will return the same value as `if server.UseTLS`(I mean, `myServ.dc1` could return UseTLS=true while `getAreaForPeerDatacenterUseTLS("dc1")` could return false. That is why I had to introduce `dc` to avoid changing too much the behaviour.

I am also quite afraid of activating silently TLS for all Consul users in such case, because for now, having `verify_outgoing` did actually disabled TLS by default over RPC (which might have a significant performance impact).

What do you think?

**Comment by dnephin** [comment]
I agree we shouldn't change the existing behaviour for normal RPC, but I think we can make the default behaviour for gRPC requests do the right thing. If someone doesn't want TLS at all, they omit the `ca_file`.  I think since the existing callers already call `UseTLS` before calling `OutgoingRPCWrapper` it should be safe to change `OutgoingRPCWrapper`.

**Comment by pierresouchay** [comment]
@dnephin what would you recommend then? Because it would be 2 different code paths, so, I would have to duplicate `OutgoingRPCWrapper`, one for GRPC over RPC, the other for RPC.

Still I don't really get the purpose of `getAreaForPeerDatacenterUseTLS()` in the codepath...

I am a bit skeptical: having different behaviour for GRPC OverRPC and other RPC is weird for the same parameters. TLS is hard enough/scary for admins, I would not complexity furthermore the behaviour if possible.

I see and applause the point of using TLS as much as possible, but I would rather change this behaviour for *all RPC*, not only GRPC over RPC in the next major release in such case.

**Comment by dnephin** [comment]
Ya, I think you are probably right. We should consider doing that for 1.10 but we don't need to do it in this PR.

Another option would then be to change the `grpc/client` to accept the full `tlsutil.Configurator` instead of only the wrapper. That way we could call `if server.UseTLS && configurator.UseTLS(dc)`, which is basically what the other code paths do.

**Comment by pierresouchay** [comment]
You decide, tell me what to do, I'll do the change quickly because it blocks our release of 1.9.x in production (we try to push changes in prod only when merged upstream)

**Comment by pierresouchay** [comment]
@dnephin Implemented as you suggested in https://github.com/hashicorp/consul/pull/9512

**Comment by dnephin** [comment]
replaced by #9512

**Pull request closed without merging**